src/persist-client/src/lib.rs

```diff
@@ -0,0 +1,315 @@
+//   we should think about how this will be used.
+// - Leases for read and write handles, as described in formalism.
+// - I think some things (readers at least?) will want to connect to the persist
+//   controller over rpc. At the very least, this is something that we may want
+//   to hook up for latency optimizations (making inputs to the state machine
+//   push instead of pull). How does this information get plumbed around? Maybe
+//   a dns name that cloud is responsible for making sure points at the right
+//   thing?
+// - To what extent should this adopt the naming conventions for STORAGE used in
+//   the formalism doc? Many of them have direct analogues down at the persist
+//   level.
+// - At the moment, we have both a STORAGE collection concept as well as a
+//   persist collection concept. One of these should be renamed :).
+
+/// A location in s3, other cloud storage, or otherwise "durable storage" used
+/// by persist. This location can contain any number of persist collections.
+///
+/// WIP: This also needs etcd (or whatever we pick) connection information.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct Location {
+    bucket: String,
+    prefix: String,
+}
+
+/// An opaque identifier for a persist collection.
+#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
+pub struct Id([u8; 16]);
+
+// WIP: Define the semantics of this. In particular, we should use this
+// opportunity to think through where we need to push retries into persist and
+// which places need structured information back about errors we can't recover
+// from.
+#[derive(Debug)]
+pub struct Error;
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        todo!("{:?}", f.width())
+    }
+}
+
+impl std::error::Error for Error {}
+
+pub trait DurableTimestamp: Timestamp {
+    // WIP: Figure out the requirements here. E.g. if we can require timestamps
+    // be roundtrip-able as 8 bytes, persist can do some nice performance things
+    // in it's internal columnar storage format.
+    //
+    // WIP: An alternative is to separate out this encode and decode into a
+    // Codec64 trait and require only the necessary Timestamp vs Codec64 bounds
+    // in various places.
+    fn encode(&self) -> [u8; 8];
+    fn decode(buf: [u8; 8]) -> Self;
+}
+
+pub trait DurableDiff: Semigroup {
+    // WIP: Figure out the requirements here. E.g. if we can require timestamps
+    // be roundtrip-able as 8 bytes, persist can do some nice performance things
+    // in it's internal columnar storage format.
+    //
+    // WIP: An alternative is to separate out this encode and decode into a
+    // Codec64 trait and require only the necessary Semigroup vs Codec64 bounds
+    // in various places.
+    fn encode(&self) -> [u8; 8];
+    fn decode(buf: [u8; 8]) -> Self;
+}
+
+pub struct Client {
+    _phantom: PhantomData<()>,
+}
+
+impl Client {
```

Comment:
Notable departures from the current persist API (for explicit discussion):
- Write and Seal are replaced by WriteBatch. This potentially makes things much harder to use multi-threaded, but makes some other things much easier. It seems to be the way the wind is blowing.
- AllowCompaction is replaced by read capabilities.
- Persist collections no longer have String external names, only an opaque identifier assigned at creation time.
- Multiple concurrent writers are supported for a single collection in the way outlined in Petros's doc.
- Only an "async" API is offered.
- Persist no longer requires write exclusivity. The normal deployment of the "controller" is that one is running, but it's safe to start a new one while the old one is in an unknown state (distributed systems amirite). In exchange, persist depends on a TBD distributed consensus service (etcd, cockroachdb, some aws thing).